pkg/capabilities/capabilities.go

```diff
@@ -1,88 +1,290 @@
+const pkgName = "capabilities"
+
+//
+// "Effective" might be at protection rings 0,1,2,3
+// "Permitted" is always at ring0 (so effective can migrate rings)
+// "Bound" will bet set to unprivileged so exec() can't inherit capabilities.
+//
+
+type ringType int
+
+const (
+	Privileged   ringType = iota // ring0 (all capabilities enabled, startup/shutdown)
+	Required                     // ring1 (needed capabilities only: config time)
+	Requested                    // ring2 (temporary specific capabilities)
+	Unprivileged                 // ring3 (no capabilities: runtime)
+)
+
+type Capabilities struct {
+	have   *cap.Set
+	all    map[cap.Value]map[ringType]bool
+	bypass bool
+	lock   *sync.Mutex // RW big lock for outer & inner maps (simpler)
+}
+
+func NewCapabilities(bypass bool) *Capabilities {
+	c := &Capabilities{}
+	c.initialize(bypass)
+	return c
+}
+
+func (c *Capabilities) initialize(bypass bool) {
+	if bypass {
+		c.bypass = true
+		return
+	}
+
+	c.lock = new(sync.Mutex)
+	c.all = make(map[cap.Value]map[ringType]bool)
+
+	for v := cap.Value(0); v < cap.MaxBits(); v++ {
+		c.all[v] = make(map[ringType]bool)
+		c.all[v][Privileged] = true // all capabilities are enabled
+		c.all[v][Required] = false
+		c.all[v][Requested] = false
+		c.all[v][Unprivileged] = false
+	}
+
+	c.getProc()
+	for c := range c.all {
+		cap.DropBound(c) // drop all capabilities from bound
+	}
+	c.setProc()
+
+	// The base for required capabilities (ring1) depends on the following:
+
+	c.Require(
+		cap.IPC_LOCK,
+		cap.SYS_RESOURCE,
+	)
+
+	// Kernels bellow v5.8 do not support cap.BPF + cap.PERFMON (instead of
+	// having to have cap.SYS_ADMIN), nevertheless, some kernels, like RHEL8
+	// clones, have backported cap.BPF capability and might be able to use it.
+
+	paranoid, err := getKernelPerfEventParanoidValue()
 	if err != nil {
-		return err
+		logger.Debug("could not get perf_event_paranoid, assuming highest", "pkg", pkgName)
 	}
-	return checkRequired(selfCaps, reqCaps)
+
+	if paranoid > 2 {
+		logger.Debug("paranoid: Value in /proc/sys/kernel/perf_event_paranoid is > 2", "pkg", pkgName)
+		logger.Debug("paranoid: Tracee needs CAP_SYS_ADMIN instead of CAP_BPF + CAP_PERFMON", "pkg", pkgName)
+		logger.Debug("paranoid: To change that behavior set perf_event_paranoid to 2 or less.", "pkg", pkgName)
+		c.Require(cap.SYS_ADMIN)
+	}
+
+	hasBPF, _ := c.have.GetFlag(cap.Permitted, cap.BPF)
+	if hasBPF {
+		c.Require(
+			cap.BPF,
+			cap.PERFMON,
+		)
+	} else {
+		c.Require(
+			cap.SYS_ADMIN,
+		)
+	}
+
+	c.Unprivileged() // default ring
 }
 
-func checkRequired(caps set, reqCaps []cap.Value) error {
-	var missingCaps []cap.Value
-	for _, c := range reqCaps {
-		exist, err := caps.GetFlag(cap.Effective, c)
-		if err != nil {
-			return err
-		}
-		if !exist {
-			missingCaps = append(missingCaps, c)
-		}
+// Public Methods
+
+// Privileged is a protection ring with all caps set as Effective.
+func (c *Capabilities) Privileged(cb func() error) error {
```

Comment:
We should. I missed adding because of a context switch (likely) and never came back to adding it there. I believe you are right regarding libcap behavior!!! Lets do it!

Reply:
Im fixing this.